graph/iterator/resolver.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2018 The Cayley Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iterator
+
+import (
+	"context"
+	"fmt"
+	"github.com/cayleygraph/cayley/graph"
+	"github.com/cayleygraph/cayley/quad"
+)
+
+var _ graph.Iterator = &Resolver{}
+
+// A Resolver iterator consists of it's order, an index (where it is in the,
+// process of iterating) and a store to resolve values from.
+type Resolver struct {
+	qs     graph.QuadStore
+	uid    uint64
+	tags   graph.Tagger
+	order  []quad.Value
+	values map[quad.Value]graph.Value
+	nodes  map[graph.Value]quad.Value
```

Comment:
It's not safe to use `graph.Value` as a key. See `Key()` method on `graph.Value`.